### ISSUE TITLE
Unable to connect to any website when socks5+ proxy type is specified

### ISSUE THREAD

**Issue by netms7** [3proxy/3proxy#712]
Hello!
I can't connect to any website when proxy type socks5+ is specified.
My config file:
```
flush
auth iponly
fakeresolve
internal 127.0.0.1

log
logformat "L%t.%. %N.%p %E %U %C:%c %R:%r %O %I %h %T"

allow *
parent 1000 socks5+ 127.0.0.1 5000

socks -p9999
```

I tested it both with fakeresolve and without, no difference,
I used libcurl to test requests.
This request return this error. There is even no hostname, just IP address.
```
./curl.exe --socks5 127.0.0.1:9999 http://1.1.1.1
curl: (1) Received HTTP/0.9 when not allowed
```
The only log line for this request:
`1637715723.544 SOCK5.9999 00095 - 127.0.0.1:53595 127.0.0.1:5000 71 416 1 CONNECT 1.1.1.1:80`
When HTTPS is used it even doesn't return any response, just infinite waiting,
Changing the proxy type from socks5+ to socks5 solves these problems and both requests get working.

The proxy server at 127.0.0.1:5000 works correctly. I tested many requests through this proxy directly and it works fine for both HTTP and HTTPS, with resolution hostnames through proxy and locally, whatever.

Why socks5+ proxy type doesn't work?
Is it a bug or I've mistaken?
Hope for your help!

**Comment by z3APA3A** [comment]
00095 means parent proxy have not accepted request. Probably, you have checked it with curl --socks5 while equivalent for socks5+ in curl is --socks5h or --socks5-hostname

**Comment by netms7** [comment]
> 00095 means parent proxy have not accepted request. Probably, you have checked it with curl --socks5 while equivalent for socks5+ in curl is --socks5h or --socks5-hostname

Unfortunately, with --socks5-hostname it doesn't work either. It doesn't work with firefox too

**Comment by z3APA3A** [comment]
Sorry, but I can't help you with this diagnostics. Generally, socks+ parent is known to work, it may be some issue specific to your parent proxy, try to check parent proxy log to identify it or may be you use some old 3proxy version, use the current one.

**Comment by netms7** [comment]
> Sorry, but I can't help you with this diagnostics. Generally, socks+ parent is known to work in current version, it may be some issue specific to your parent proxy, try to check parent proxy log to identify it or may be you use some old 3proxy version, use the current one.

I've analyzed curl and 3proxy requests with wireshark.
The only difference between these requests that 3proxy provides only No Authentication(0x00) method, while libcurl provides No Authentication(0x00) and GSSAPI(0x01) methods. And when resolving hostname through proxy (Address Type is Domain Name(0x03)) the actual proxy server (127.0.0.1:5001) replies ATYP Domain Name, Remote Address 1.1.1.1 and port 80 (As specified in request packet), but 3proxy replies to the client ATYP IPv4, Remote Address 127.0.0.1 and Port is the client connection port + 1.
But anyway the actual proxy server returns response from 1.1.1.1 and 3proxy closes connection with 127.0.0.1:5001 normally, but 3proxy anyway doesn't return the response to the client.
I've attached two packet captures from wireshark: direct request to the actual proxy with libcurl and request from libcurl through 3proxy to the actual proxy.
You can analyze it by yourself.

The actual proxy server at 127.0.0.1:5001. The 3proxy SOCKS server at 127.0.0.1:7000. 
[captures.zip](https://github.com/3proxy/3proxy/files/7596535/captures.zip)

Hope you will check it! Thanks in advance!

**Comment by z3APA3A** [comment]
which 3proxy version are you using?

**Comment by z3APA3A** [comment]
There is something weird in this capture. Look for packet 23. It's IP fragment  and it contains "50" octet in TCP part. This octet comes out-of-order, breaks HTTP and leads to "HTTP 0.9" reply in curl, and is something impossible, because next IP fragment is in the packet 31. Packet 31 contains valid HTTP reply BUT packet 23 can not be a fragment of packet 31 because HTTP request comes in packet 25 _after_ packet 23 and http reply could not be sent prior to it. Packet 23 is a source of problem - this packet should not exist and it's definitely can not be inserted by 3proxy, because 3proxy does not deal with raw sockets and it's inserted on network layer. May be you have some active filter on network layer (ipfilter? ebpf?) which injects this packet.

**Comment by netms7** [comment]
> which 3proxy version are you using?

The latest released on github 0.9.4. This release: https://github.com/3proxy/3proxy/releases/tag/0.9.4

**Comment by z3APA3A** [comment]
Well, I figured out the problem, for some reason parent SOCKS replies to connect request with destination address in domain name (string) format. According to RFC 1918 this address/port should contain bind address (that is, address of proxy interface used to connect to destination server) and it's quite unexpected to be a string, it should be IPv4 or IPv6. While this behavior violates RFC, using domain name in connect reply does not (though it's unreasonable), I will support this type of reply in 3proxy, but you better report non-standard behavior to parent socks developers. Some protocols, e.g. FTP or torrent require valid interface address here.

**Comment by netms7** [comment]
Unfortunately, I don't own servers that I use and I doubt the developers of those servers will fix it. I would be grateful if you could add support for this type of reply. I really need that feature to work.

**Comment by z3APA3A** [comment]
Try current master (with bd1dcacf73f2dd1bf6bfd08d20aee39b524e3061)

**Comment by netms7** [comment]
> Try current master (with [bd1dcac](https://github.com/3proxy/3proxy/commit/bd1dcacf73f2dd1bf6bfd08d20aee39b524e3061))

Actually, it doesn't work too. But now there is another error:
`curl: (7) Can't complete SOCKS5 connection to 127.0.0.1:7001. (4)`
The only log line for this request:
`1637793102.711 SOCK5.7001 00059 - 127.0.0.1:50364 127.0.0.1:5001 0 60 2 CONNECT 1.1.1.1:80`
I've built it with MSVC compiler if it matters.
I also analyzed the traffic and found that now 3proxy closes connection just after receiving SOCKS reply from the parent proxy and return to the client Host unreachable (4) reason.
Attached traffic capture: [capture.zip](https://github.com/3proxy/3proxy/files/7598995/capture.zip)
3proxy at 7001 port, parent proxy at 5001 port.
By the way, I get this error at 3proxy start for each started proxy:
`[line 18] Warning: no nserver/nscache configured, socks may run very slow`
